File: 3961-maximize-sum-of-device-ratings/3961-maximize-sum-of-device-ratings.cpp
class Solution {
public:
    long long maxRatings(vector<vector<int>>& u) {
        long long s = 0, t = 0;
        int m = 1e6, g = 1e6, h = 1e6, k = 1e6;

        for (auto& r : u) {
            int a = 1e6, b = 1e6;

            for (int x : r)
                x < a ? (b = a, a = x) : x < b ? b = x : 0;

            int d = b < 1e6 && b > a ? b - a : 0;
            int e = a + d;

            s += a;
            t += d;

            a < m
                ? (k = min(k, h), m = a, g = d, h = e)
                : a == m
                    ? (g = min(g, d), h = min(h, e))
                    : k = min(k, e);
        }

        return s + t - min((long long)g, (long long)k - m);
    }
};

File: 3961-maximize-sum-of-device-ratings/3961-maximize-sum-of-device-ratings.java
class Solution {
    public long maxRatings(int[][] units) {
        int m = units.length;
        int n = units[0].length;
        
        if (n == 1) {
            long sum = 0;
            for (int i = 0; i < m; i++) {
                sum += units[i][0];
            }
            return sum;
        }
        
        long sumA2 = 0;
        int M = Integer.MAX_VALUE;
        int minA2 = Integer.MAX_VALUE;
        
        for (int i = 0; i < m; i++) {
            int min1 = Integer.MAX_VALUE;
            int min2 = Integer.MAX_VALUE;
            
            for (int j = 0; j < n; j++) {
                int val = units[i][j];
                if (val < min1) {
                    min2 = min1;
                    min1 = val;
                } else if (val < min2) {
                    min2 = val;
                }
            }
            
            if (min1 < M) {
                M = min1;
            }
            
            sumA2 += min2;
            
            if (min2 < minA2) {
                minA2 = min2;
            }
        }
        
        return (long) M + sumA2 - minA2;
    }
}